### PR TITLE
New listener resource for the federation API "openid/userinfo" endpoint

### PR DESCRIPTION
**Pull request by jaywink**
Integration managers use the OpenID userinfo endpoint in the federation API to verify that user OpenID access tokens are valid. If the federation resource is disabled, integration managers will not be able to verify the access token, causing a broken experience for users. The OpenID userinfo endpoint has now been split to a separate `openid` resource, which is enabled by default in newly generated configuration. It is also enabled automatically if the federation resource is enabled.

Requires no actions on homeservers which have federation listener enabled. For homeservers with federation listener disabled, the `openid` listener resource needs to be manually added to the existing configuration, if the resource is wanted to be enabled. New installations will adopt the default of activating the resource even if federation resource is disabled.

Add also `parameterized` Python module to be able to easily parameterize test runs.

### CODE REVIEW

File: changelog.d/4420.feature [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+OpenID access tokens are valid. If the federation resource is disabled, integration managers will not be able
+to verify the access token, causing a broken experience for users. The OpenID userinfo endpoint has now been split
+to a separate `openid` resource, which is enabled by default in newly generated configuration. It is also enabled
+automatically if the federation resource is enabled.
+
+If your homeserver runs federation enabled, this change does not require any actions.
+
+If you run a homeserver with federation disabled, we recommend adding the `openid` resource to your homeserver
+configuration in the `type: http` listener `resources` list to allow your users access to
+integration manager features.
```
Comment (erikjohnston) [inline changelog.d/4420.feature]: This is an *incredibly* nice change log :+1: , but alas it needs to be just the one line :cry: 

However the changelog entry will point to this PR, so people will be able to find this info from the PR description :tada:

File: synapse/config/server.py [hunk 2]
```diff
@@ -350,7 +350,7 @@ def default_config(self, server_name, data_dir_path, **kwargs):
             resources:
               - names: [client]
                 compress: true
-              - names: [federation]
+              - names: [federation, openid]
```
Comment (erikjohnston) [inline synapse/config/server.py]: I wouldn't bother adding these if they get added by `federation`. You might want to add a separate (commented out) entry to the `resources` section (like we do below for the `manhole` listener) to document that `openid` is a thing. Maybe something like:

```yaml
# # If federation is disabled synapse can still expose the open ID endpoint
# # to allow integrations to authenticate users
# - names: [openid]
#   compress
```

File: synapse/federation/transport/server.py [hunk 3]
```diff
@@ -42,9 +42,10 @@
 class TransportLayerServer(JsonResource):
     """Handles incoming federation HTTP requests"""
 
-    def __init__(self, hs):
+    def __init__(self, hs, servlet_groups=None):
```
Comment (erikjohnston) [inline synapse/federation/transport/server.py]: Can you document the `servlet_groups` param please

File: synapse/federation/transport/server.py [hunk 4]
```diff
@@ -1314,44 +1318,70 @@ def on_PUT(self, origin, content, query, group_id):
...
-        ).register(resource)
-
-    for servletclass in GROUP_ATTESTATION_SERVLET_CLASSES:
-        servletclass(
-            handler=hs.get_groups_attestation_renewer(),
-            authenticator=authenticator,
-            ratelimiter=ratelimiter,
-            server_name=hs.hostname,
-        ).register(resource)
+def register_servlets(hs, resource, authenticator, ratelimiter, servlet_groups=None):
```
Comment (erikjohnston) [inline synapse/federation/transport/server.py]: Can you document the `servlet_groups` param please

File: tests/app/test_openid_listener.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+from mock import patch, Mock
+from parameterized import parameterized
+
+from synapse.app.federation_reader import FederationReaderServer
+from synapse.app.homeserver import SynapseHomeServer
+
+from tests.unittest import HomeserverTestCase
+
+
+@patch("synapse.app.homeserver.KeyApiV2Resource", new=Mock())
```
Comment (erikjohnston) [inline tests/app/test_openid_listener.py]: Why is this necessary?

File: tests/app/test_openid_listener.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        # Grab the resource from the site that was told to listen
+        site = self.reactor.tcpServers[0][1]
+        try:
+            self.resource = (
+                site.resource.children[b"_matrix"].children[b"federation"].children[b"v1"]
+            )
+        except KeyError:
+            if expectation == "no_resource":
+                return
+            raise
```
Comment (erikjohnston) [inline tests/app/test_openid_listener.py]: I'm not sure what this is testing? Can't we just test that the below attempt at requesting fails appropriately?

**Comment by erikjohnston** [comment]
BTW, travis is failing on the `isort` tests, which is basically just checking the imports are correctly ordered/etc. You can easily fix it by running `isort -rc synapse/ tests/`.

File: changelog.d/4420.feature (hunk 1 above)
Comment (hawkowl) [reply changelog.d/4420.feature]: it doesn't have to be one line (or if we strip whitespace, we can configure it not to). the real problem is breaking the dot-pointing, which i forget how to stop with markdown.

Comment (jaywink) [reply changelog.d/4420.feature]: Will condense to one line! :) I'm a fan of verbose changelogs, but I will live ;)

@hawkowl what do you mean by this, just out of interest?

> he real problem is breaking the dot-pointing, which i forget how to stop with markdown.

File: tests/app/test_openid_listener.py (hunk 5 above)
Comment (jaywink) [reply tests/app/test_openid_listener.py]: The tests blow up otherwise due to some internals in the `KeyApiV2Resource` which would require further tweaking of the test homeserver possibly. Since we're not testing that, it was easier to mock it.

I'll post here what the problem was when refactoring the tests.

File: tests/app/test_openid_listener.py (hunk 6 above)
Comment (jaywink) [reply tests/app/test_openid_listener.py]: So, I took how the tests were written pretty much from `test_frontend_proxy`, added here https://github.com/matrix-org/synapse/pull/3694/files#diff-20dd51850158a57f8dd29117bf46b484R50. 

What I noticed is that if I don't call `self.resource = site.resource.children[b"_matrix"].children[b"federation"].children[b"v1"]` at all, then everything fails with a 400 error. But when there is no federation listener, calling this will fail due to  `KeyError`. I tried using `.get()` but that didn't have the same effect. I'm not entirely sure how the test `make_request` works underneath. This was also the same reason I introduced `parameterized` because otherwise there would have been a lot of duplication of the test code when in fact we just want to run the same test with different parameters.

I'm not entirely sure how to test a failure since we can't call `self.resource` without that failing, and if it's not called, any request fails, so it's not really a successful test. Maybe @hawkowl has some details on that as the author the frontend proxy test, what do I need to do here to make `make_request` works without calling `self.resource` first?

File: tests/app/test_openid_listener.py (hunk 5 above)
Comment (jaywink) [reply tests/app/test_openid_listener.py]:
```
  File "/home/jaywink/workspace/synapse/synapse/app/homeserver.py", line 227, in _configure_named_resource
    resources[SERVER_KEY_V2_PREFIX] = KeyApiV2Resource(self)
  File "/home/jaywink/workspace/synapse/synapse/rest/key/v2/__init__.py", line 25, in __init__
    self.putChild(b"server", LocalKey(hs))
  File "/home/jaywink/workspace/synapse/synapse/rest/key/v2/local_key_resource.py", line 70, in __init__
    self.update_response_body(self.clock.time_msec())
  File "/home/jaywink/workspace/synapse/synapse/rest/key/v2/local_key_resource.py", line 75, in update_response_body
    self.valid_until_ts = int(time_now_msec + refresh_interval)
TypeError: unsupported operand type(s) for +: 'int' and 'Mock'
```

Just noticed it's only a problem with `SynapseHomeserverOpenIDListenerTests` though, not the reader. If federation resource is active, key resource is added automatically, thus the mock.

File: tests/app/test_openid_listener.py (hunk 6 above)
Comment (erikjohnston) [reply tests/app/test_openid_listener.py]: Oh, I hadn't realised we had done this somewhere else. Will poke @hawkowl to see if there's a better way of doing it, but if not I can live with the current stuff

**Comment by erikjohnston** [comment]
Oh, oops, I think all the test failures are my fault as I changed the listener to listen on `/_matrix/federation/` rather than `/_matrix/federation/v1`, but hopefully an easy fix by just removing the `.children["v1"]` bit

**Comment by jaywink** [comment]
Rebased and fixed the tests.

**Comment by erikjohnston** [comment]
I'm going to merge this as the code looks correct and there are tests. I feel like there should be a neater way of doing the tests, but if they're using the same style as other unit tests its fine for now.
